webapi/src/main/scala/org/knora/webapi/slice/ontology/api/service/RestCardinalityService.scala

```diff
@@ -0,0 +1,122 @@
+      case (Some(_), None)                           => ZIO.fail(missingQueryParamValue(newCardinalityKey))
+      case (None, None)                              => canReplaceCardinality(classIri, user)
+      case (Some(propertyIri), Some(newCardinality)) => canSetCardinality(classIri, propertyIri, newCardinality, user)
+    }
+
+  def canReplaceCardinality(classIri: IRI, user: UserADM): Task[CanDoResponseV2]
+
+  def canSetCardinality(
+    classIri: IRI,
+    propertyIri: IRI,
+    cardinality: String,
+    user: UserADM
+  ): Task[CanDoResponseV2]
+}
+
+private final case class PermissionService(ontologyRepo: OntologyRepo) {
+  def hasOntologyWriteAccess(user: UserADM, ontologyIri: InternalIri): Task[Boolean] = {
+    val permissions = user.permissions
+    for {
+      data           <- ontologyRepo.findById(ontologyIri)
+      projectIriMaybe = data.flatMap(_.ontologyMetadata.projectIri.map(_.toIri))
+      hasPermission   = projectIriMaybe.exists(permissions.isSystemAdmin || permissions.isProjectAdmin(_))
+    } yield hasPermission
+  }
+}
+
+case class RestCardinalityServiceLive(
+  cardinalityService: CardinalityService,
+  iriConverter: IriConverter,
+  ontologyRepo: OntologyRepo
+) extends RestCardinalityService {
+
+  private val permissionService: PermissionService = PermissionService(ontologyRepo)
+
+  def canReplaceCardinality(classIri: IRI, user: UserADM): Task[CanDoResponseV2] =
+    for {
+      classIri <- iriConverter.asInternalIri(classIri).orElseFail(invalidQueryParamValue(classIriKey))
+      _        <- checkUserHasWriteAccessToOntologyOfClass(user, classIri)
+      result   <- cardinalityService.canReplaceCardinality(classIri)
+    } yield toResponse(result)
+
+  private def checkUserHasWriteAccessToOntologyOfClass(user: UserADM, classIri: InternalIri): Task[Unit] = {
+    val hasWriteAccess = for {
+      ontologyIri    <- iriConverter.getOntologyIriFromClassIri(classIri)
+      hasWriteAccess <- permissionService.hasOntologyWriteAccess(user, ontologyIri)
+    } yield hasWriteAccess
+    ZIO.ifZIO(hasWriteAccess)(
+      onTrue = ZIO.unit,
+      onFalse = ZIO.fail(ForbiddenException(s"User has no write access to ontology"))
+    )
+  }
+
+  def canSetCardinality(
+    classIri: IRI,
+    propertyIri: IRI,
+    cardinality: String,
+    user: UserADM
+  ): Task[CanDoResponseV2] =
+    for {
+      classIri       <- iriConverter.asInternalIri(classIri).orElseFail(invalidQueryParamValue(classIriKey))
+      _              <- checkUserHasWriteAccessToOntologyOfClass(user, classIri)
+      newCardinality <- parseCardinality(cardinality).orElseFail(invalidQueryParamValue(newCardinalityKey))
+      propertyIri    <- iriConverter.asInternalIri(propertyIri).orElseFail(invalidQueryParamValue(propertyIriKey))
+      result         <- cardinalityService.canSetCardinality(classIri, propertyIri, newCardinality)
+    } yield toResponse(result)
+
+  private def parseCardinality(cardinality: String): IO[Option[Nothing], Cardinality] =
```

Comment:
```suggestion
  private def parseCardinalityFromString(cardinality: String): IO[Option[Nothing], Cardinality] =
```

Reply:
I would like to keep it as is as the type of the argument is already clear enough IMHO.